File: iris-macos/leanring-buddy/UsageMonitor.swift
//
//  UsageMonitor.swift
//  leanring-buddy
//
//  Anonymous usage counts: which catalog apps are opened, which AI provider
//  and tier answer, which guides start and finish. Founder, 2026-09-25:
//  "depersonalize the tracking, make it optional, prompt them when they open
//  it, default toggled on, must not break anything in Iris, nothing should
//  wait for it."
//
//  WHAT AN EVENT CAN HOLD. A catalog slug, one of five event kinds, a provider
//  and a tier as enum values, the OS family, the Iris version, and the HOUR it
//  happened in. That is the whole vocabulary: there is no field for a message,
//  a window title, a path, a model name or an account, so no call site can
//  pass one by accident. The server (`lib/iris-usage-events.ts` in publik)
//  refuses any other field and drops a slug its catalog does not list.
//
//  NOTHING WAITS FOR IT. `record` hops onto a private utility queue and
//  returns. Events are held in memory as counts per (event, hour); a timer on
//  the same queue sends one batch at most every 60 seconds. A send that fails
//  for any reason — offline, a 4xx, a 5xx, a timeout — is DROPPED, never
//  retried and never written to disk. While a send is in flight the next tick
//  is skipped. Nothing in this file is awaited by the UI or by any flow.
//
//  OFF MEANS OFF. The switch is read at record time and again at send time.
//  With it off, nothing is recorded and whatever was held is thrown away.
//

import Foundation

/// The five things that are counted. Raw values are the wire names.
nonisolated enum UsageEventKind: String, Sendable, CaseIterable {
    case appOpened = "app_opened"
    case aiCall = "ai_call"
    case modelSelected = "model_selected"
    case guideStarted = "guide_started"
    case guideCompleted = "guide_completed"
}

/// Which route answered, as the server's enum. Mirrors
/// `AssistantProviderPreference`'s raw values, kept separate so this file has
/// no main-actor dependency.
nonisolated enum UsageProvider: String, Sendable, CaseIterable {
    case publikAPI = "publik-api"
    case anthropicKey = "anthropic-key"
    case codex = "codex"
}

/// publik's three tiers. Iris's picker speaks in Anthropic model names; this
/// is the same coarse mapping `PublikAPIModelAlias` uses, so the count and the
/// gateway agree about what "the tier" was.
nonisolated enum UsageModelTier: String, Sendable, CaseIterable {
    case fast
    case balanced
    case smart

    static func forIrisModelName(_ irisModelName: String) -> UsageModelTier {
        let lowercasedName = irisModelName.lowercased()
        if lowercasedName.contains("haiku") { return .fast }
        if lowercasedName.contains("opus") { return .smart }
        return .balanced
    }
}

/// One thing that happened. Built only from enums and a catalog slug.
nonisolated struct UsageEvent: Hashable, Sendable {
    let kind: UsageEventKind
    let appSlug: String?
    let provider: UsageProvider?
    let modelTier: UsageModelTier?

    init(kind: UsageEventKind, appSlug: String? = nil, provider: UsageProvider? = nil, modelTier: UsageModelTier? = nil) {
        self.kind = kind
        // A slug that is not slug-shaped is dropped here rather than sent and
        // refused: the server would reject the whole batch over it.
        self.appSlug = appSlug.flatMap { UsageEvent.isSlugShaped($0) ? $0 : nil }
        self.provider = provider
        self.modelTier = modelTier
    }

    /// The server's slug rule (`apps.slug`): lowercase letters, digits and
    /// inner hyphens, 1–64 characters.
    static func isSlugShaped(_ candidate: String) -> Bool {
        candidate.range(of: "^[a-z0-9]+(?:[a-z0-9-]{0,62}[a-z0-9])?$", options: .regularExpression) != nil
    }
}

/// Where a batch goes. The real one is `URLSessionUsageEventSender`; tests
/// substitute a recorder.
nonisolated protocol UsageEventSending: Sendable {
    /// Send one batch. `completion` is called exactly once with whether the
    /// server accepted it; the monitor drops the batch either way.
    func sendBatch(_ jsonBody: Data, completion: @escaping @Sendable (Bool) -> Void)
    /// Ask the server to erase every count this install sent (the switch was
    /// turned off). Fire and forget.
    func eraseEverythingSent(byInstallIdentifier installIdentifier: String)
}

nonisolated final class UsageMonitor: @unchecked Sendable {

    struct Configuration: Sendable {
        /// Read at record time and at send time.
        let isSharingEnabled: @Sendable () -> Bool
        /// The consent.json install id.
        let installIdentifier: @Sendable () -> String
        /// "0.9.15". A version that is not three numbers is padded; one that
        /// cannot be read at all turns sending off, since the server would
        /// refuse every batch.
        let irisVersion: String
        /// "macos" or "windows".
        let operatingSystem: String
        let sender: UsageEventSending
        let currentDate: @Sendable () -> Date
        /// At most one send per this many seconds.
        let flushInterval: TimeInterval
        /// How many distinct (event, hour) counts are held between sends.
        /// Anything new past this is dropped until the next send.
        let maximumDistinctCountsHeld: Int
    }

    /// Past this the server refuses a row (`MAX_COUNT_PER_EVENT`).
    static let largestCountTheServerAccepts = 500
    /// The server's batch cap (`MAX_USAGE_EVENTS_PER_BATCH`).
    static let largestBatchTheServerAccepts = 50

    /// The app-wide monitor. Inert until `configure` has run, so a call site
    /// that fires before launch finishes simply records nothing.
    static let shared = UsageMonitor()

    private struct HeldCountKey: Hashable {
        let event: UsageEvent
        let hourBucket: String
    }

    private let workQueue: DispatchQueue
    // Everything below is touched only on `workQueue`.
    private var configuration: Configuration?
    private var heldCounts: [HeldCountKey: Int] = [:]
    /// Arrival order, so a batch is sent oldest-first and a truncated batch
    /// drops the newest rather than an arbitrary set.
    private var heldCountOrder: [HeldCountKey] = []
    private var aSendIsInFlight = false
    private var flushTimer: DispatchSourceTimer?
    /// How many events were thrown away (cap, off, failed send). For tests
    /// and for the log line; never sent anywhere.
    private var droppedEventCount = 0

    init(workQueue: DispatchQueue = DispatchQueue(label: "com.publikhq.iris.usage-monitor", qos: .utility)) {
        self.workQueue = workQueue
    }

    // MARK: - Setup

    func configure(_ configuration: Configuration) {
        workQueue.async { self.configuration = configuration }
    }

    /// Starts the send timer. Safe to call more than once.
    func start() {
        workQueue.async {
            guard self.flushTimer == nil, let configuration = self.configuration else { return }
            let timer = DispatchSource.makeTimerSource(queue: self.workQueue)
            timer.schedule(
                deadline: .now() + configuration.flushInterval,
                repeating: configuration.flushInterval,
                leeway: .seconds(5)
            )
            timer.setEventHandler { [weak self] in self?.sendWhatIsHeld() }
            self.flushTimer = timer
            timer.resume()
        }
    }

    func stop() {
        workQueue.async {
            self.flushTimer?.cancel()
            self.flushTimer = nil
        }
    }

    // MARK: - Recording

    /// Counts one event. Returns immediately; never throws; never blocks.
    func record(_ event: UsageEvent) {
        workQueue.async {
            guard let configuration = self.configuration else { return }
            guard configuration.isSharingEnabled() else {
                self.throwAwayEverythingHeld()
                return
            }
            let key = HeldCountKey(
                event: event,
                hourBucket: Self.hourBucket(for: configuration.currentDate())
            )
            if let existingCount = self.heldCounts[key] {
                self.heldCounts[key] = min(existingCount + 1, Self.largestCountTheServerAccepts)
                return
            }
            guard self.heldCounts.count < configuration.maximumDistinctCountsHeld else {
                self.droppedEventCount += 1
                return
            }
            self.heldCounts[key] = 1
            self.heldCountOrder.append(key)
        }
    }

    /// The switch was turned off: forget what is held, and ask the server to
    /// forget what it has.
    func sharingWasTurnedOff() {
        workQueue.async {
            self.throwAwayEverythingHeld()
            guard let configuration = self.configuration else { return }
            configuration.sender.eraseEverythingSent(byInstallIdentifier: configuration.installIdentifier())
        }
    }

    // MARK: - Sending

    private func sendWhatIsHeld() {
        guard let configuration else { return }
        guard configuration.isSharingEnabled() else {
            throwAwayEverythingHeld()
            return
        }
        guard !aSendIsInFlight, !heldCountOrder.isEmpty else { return }
        guard let irisVersion = Self.threePartVersion(configuration.irisVersion) else {
            throwAwayEverythingHeld()
            return
        }

        let keysToSend = Array(heldCountOrder.prefix(Self.largestBatchTheServerAccepts))
        let events: [[String: Any]] = keysToSend.map { key in
            var entry: [String: Any] = [
                "event": key.event.kind.rawValue,
                "os": configuration.operatingSystem,
                "irisVersion": irisVersion,
                "hourBucket": key.hourBucket,
                "count": heldCounts[key] ?? 1,
            ]
            if let appSlug = key.event.appSlug { entry["appSlug"] = appSlug }
            if let provider = key.event.provider { entry["provider"] = provider.rawValue }
            if let modelTier = key.event.modelTier { entry["modelTier"] = modelTier.rawValue }
            return entry
        }
        // Whatever did not fit in this batch is dropped with it: the next
        // minute starts from nothing rather than growing a backlog.
        droppedEventCount += max(0, heldCountOrder.count - keysToSend.count)
        heldCounts.removeAll()
        heldCountOrder.removeAll()

        let body: [String: Any] = ["installId": configuration.installIdentifier(), "events": events]
        guard let jsonBody = try? JSONSerialization.data(withJSONObject: body, options: [.sortedKeys]) else { return }

        aSendIsInFlight = true
        configuration.sender.sendBatch(jsonBody) { [weak self] _ in
            // Accepted or not, the batch is gone. See the file note.
            self?.workQueue.async { self?.aSendIsInFlight = false }
        }
    }

    private func throwAwayEverythingHeld() {
        droppedEventCount += heldCounts.values.reduce(0, +)
        heldCounts.removeAll()
        heldCountOrder.removeAll()
    }

    // MARK: - Wire format helpers

    /// The hour an event happened in, UTC: "2026-09-25T14:00:00Z".
    static func hourBucket(for date: Date) -> String {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(identifier: "UTC")!
        let components = calendar.dateComponents([.year, .month, .day, .hour], from: date)
        return String(
            format: "%04d-%02d-%02dT%02d:00:00Z",
            components.year ?? 1970, components.month ?? 1, components.day ?? 1, components.hour ?? 0
        )
    }

    /// "0.9" → "0.9.0", "0.9.15" → itself, "1.2.3.4" → "1.2.3", "dev" → nil.
    static func threePartVersion(_ version: String) -> String? {
        let numericParts = version.split(separator: ".").map(String.init)
        guard !numericParts.isEmpty,
              numericParts.allSatisfy({ !$0.isEmpty && $0.count <= 4 && $0.allSatisfy(\.isNumber) }) else {
            return nil
        }
        let padded = (numericParts + ["0", "0", "0"]).prefix(3)
        return padded.joined(separator: ".")
    }

    // MARK: - For tests

    /// Runs one send tick now, on the monitor's own queue, and waits for it.
    func sendNowAndWaitForTesting() {
        workQueue.sync { self.sendWhatIsHeld() }
    }

    /// Waits for every queued `record` to land.
    func drainForTesting() {
        workQueue.sync {}
    }

    var heldCountsForTesting: [UsageEvent: Int] {
        workQueue.sync {
            var byEvent: [UsageEvent: Int] = [:]
            for (key, count) in heldCounts { byEvent[key.event, default: 0] += count }
            return byEvent
        }
    }

    var droppedEventCountForTesting: Int { workQueue.sync { droppedEventCount } }
    var aSendIsInFlightForTesting: Bool { workQueue.sync { aSendIsInFlight } }
}

// MARK: - The real sender

/// POSTs a batch to `{publik}/api/telemetry/usage`. Ephemeral session: no
/// cookies, no cache, no credential of any kind — a usage count has no
/// account, so nothing that could identify one is attached.
nonisolated struct URLSessionUsageEventSender: UsageEventSending {
    let endpoint: URL
    let session: URLSession

    init(publikBaseURL: URL) {
        self.endpoint = publikBaseURL.appendingPathComponent("api/telemetry/usage")
        let configuration = URLSessionConfiguration.ephemeral
        configuration.timeoutIntervalForRequest = 10
        configuration.timeoutIntervalForResource = 15
        configuration.httpCookieStorage = nil
        configuration.urlCache = nil
        configuration.waitsForConnectivity = false
        self.session = URLSession(configuration: configuration)
    }

    func sendBatch(_ jsonBody: Data, completion: @escaping @Sendable (Bool) -> Void) {
        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = jsonBody
        session.dataTask(with: request) { _, response, error in
            let statusCode = (response as? HTTPURLResponse)?.statusCode ?? 0
            completion(error == nil && (200..<300).contains(statusCode))
        }.resume()
    }

    func eraseEverythingSent(byInstallIdentifier installIdentifier: String) {
        guard var components = URLComponents(url: endpoint, resolvingAgainstBaseURL: false) else { return }
        components.queryItems = [URLQueryItem(name: "install_id", value: installIdentifier)]
        guard let url = components.url else { return }
        var request = URLRequest(url: url)
        request.httpMethod = "DELETE"
        session.dataTask(with: request) { _, _, _ in }.resume()
    }
}

File: iris-windows/src/services/usage-monitor.ts
/**
 * Anonymous usage counts — the Windows half of `UsageMonitor.swift`.
 *
 * Founder, 2026-09-25: "depersonalize the tracking, make it optional, prompt
 * them when they open it, default toggled on, must not break anything in Iris,
 * nothing should wait for it."
 *
 * WHAT AN EVENT CAN HOLD. A catalog slug, one of five event kinds, a provider
 * and a tier as enum values, the OS family, the Iris version, and the HOUR it
 * happened in. There is no field for a message, a window title, a path, a
 * model name or an account, so no call site can pass one by accident. The
 * server (`lib/iris-usage-events.ts` in publik) refuses any other field and
 * drops a slug its catalog does not list.
 *
 * NOTHING WAITS FOR IT. `record` updates an in-memory count and returns. A
 * timer sends one batch at most every 60 seconds; a send is never awaited by
 * anything, and ANY failure — offline, a 4xx, a 5xx, a timeout, a throw —
 * drops the batch. Nothing is retried and nothing is written to disk.
 *
 * OFF MEANS OFF. The switch is read at record time and again at send time.
 * With it off nothing is recorded and whatever was held is thrown away.
 *
 * Pure: the clock, the switch, the install id and the network are injected,
 * so the whole thing runs in vitest.
 */

export const USAGE_EVENT_KINDS = [
  "app_opened",
  "ai_call",
  "model_selected",
  "guide_started",
  "guide_completed",
] as const;
export type UsageEventKind = (typeof USAGE_EVENT_KINDS)[number];

/** Which route answered. Same wire values as macOS and the server. */
export type UsageProvider = "publik-api" | "anthropic-key" | "codex";

/** publik's three tiers. */
export type UsageModelTier = "fast" | "balanced" | "smart";

export interface UsageEvent {
  kind: UsageEventKind;
  appSlug?: string | null;
  provider?: UsageProvider | null;
  modelTier?: UsageModelTier | null;
}

/** The server's cap on one row (`MAX_COUNT_PER_EVENT`). */
export const LARGEST_COUNT_THE_SERVER_ACCEPTS = 500;
/** The server's batch cap (`MAX_USAGE_EVENTS_PER_BATCH`). */
export const LARGEST_BATCH_THE_SERVER_ACCEPTS = 50;
/** Distinct (event, hour) counts held between sends; anything new past this is dropped. */
export const MAXIMUM_DISTINCT_COUNTS_HELD = 200;
/** At most one send per this long. */
export const USAGE_FLUSH_INTERVAL_MS = 60_000;

const SLUG_PATTERN = /^[a-z0-9]+(?:[a-z0-9-]{0,62}[a-z0-9])?$/;

/** The server's slug rule (`apps.slug`). */
export function isSlugShaped(candidate: string): boolean {
  return SLUG_PATTERN.test(candidate);
}

/**
 * The same coarse mapping the gateway alias uses on macOS
 * (`PublikAPIModelAlias`): haiku → fast, opus → smart, anything else →
 * balanced.
 */
export function modelTierForModelName(modelName: string): UsageModelTier {
  const lowercasedName = modelName.toLowerCase();
  if (lowercasedName.includes("haiku")) return "fast";
  if (lowercasedName.includes("opus")) return "smart";
  return "balanced";
}

/** A transport tier (`services/assistant-transport.ts`) → the wire enum. */
export function usageProviderForTransportTier(tier: "publik" | "byo" | "codex"): UsageProvider {
  if (tier === "publik") return "publik-api";
  if (tier === "byo") return "anthropic-key";
  return "codex";
}

/** The hour an event happened in, UTC: "2026-09-25T14:00:00Z". */
export function hourBucket(date: Date): string {
  const truncated = new Date(date.getTime());
  truncated.setUTCMinutes(0, 0, 0);
  return truncated.toISOString().replace(".000Z", "Z");
}

/** "0.9" → "0.9.0", "0.9.15" → itself, "1.2.3.4" → "1.2.3", "dev" → null. */
export function threePartVersion(version: string): string | null {
  const parts = version.split(".");
  if (parts.length === 0 || !parts.every((part) => /^\d{1,4}$/.test(part))) return null;
  return [...parts, "0", "0", "0"].slice(0, 3).join(".");
}

export interface UsageMonitorOptions {
  isSharingEnabled: () => boolean;
  installIdentifier: () => string;
  irisVersion: string;
  operatingSystem: "windows" | "macos";
  /** Resolves true when the server accepted the batch. Never awaited by a caller of `record`. */
  sendBatch: (jsonBody: string) => Promise<boolean>;
  /** The switch was turned off: ask the server to forget this install's counts. */
  eraseEverythingSent: (installIdentifier: string) => void;
  now?: () => Date;
  maximumDistinctCountsHeld?: number;
}

interface HeldCount {
  event: UsageEvent;
  hourBucket: string;
  count: number;
}

export class UsageMonitor {
  private readonly options: UsageMonitorOptions;
  private readonly now: () => Date;
  private readonly maximumDistinctCountsHeld: number;
  /** Insertion-ordered, so a batch goes oldest-first and a truncated one drops the newest. */
  private readonly heldCounts = new Map<string, HeldCount>();
  private aSendIsInFlight = false;
  private droppedEventCount = 0;
  private flushTimer: ReturnType<typeof setInterval> | undefined;

  constructor(options: UsageMonitorOptions) {
    this.options = options;
    this.now = options.now ?? (() => new Date());
    this.maximumDistinctCountsHeld = options.maximumDistinctCountsHeld ?? MAXIMUM_DISTINCT_COUNTS_HELD;
  }

  /** Counts one event. Synchronous, O(1), never throws. */
  record(event: UsageEvent): void {
    try {
      if (!this.options.isSharingEnabled()) {
        this.throwAwayEverythingHeld();
        return;
      }
      const appSlug = event.appSlug && isSlugShaped(event.appSlug) ? event.appSlug : null;
      const normalizedEvent: UsageEvent = {
        kind: event.kind,
        appSlug,
        provider: event.provider ?? null,
        modelTier: event.modelTier ?? null,
      };
      const bucket = hourBucket(this.now());
      const key = [normalizedEvent.kind, appSlug ?? "", normalizedEvent.provider ?? "", normalizedEvent.modelTier ?? "", bucket].join("|");
      const existing = this.heldCounts.get(key);
      if (existing) {
        existing.count = Math.min(existing.count + 1, LARGEST_COUNT_THE_SERVER_ACCEPTS);
        return;
      }
      if (this.heldCounts.size >= this.maximumDistinctCountsHeld) {
        this.droppedEventCount += 1;
        return;
      }
      this.heldCounts.set(key, { event: normalizedEvent, hourBucket: bucket, count: 1 });
    } catch {
      // Counting is never allowed to break the thing being counted.
    }
  }

  /** Starts the once-a-minute send. Safe to call more than once. */
  start(intervalMs: number = USAGE_FLUSH_INTERVAL_MS): void {
    if (this.flushTimer !== undefined) return;
    this.flushTimer = setInterval(() => this.sendWhatIsHeld(), intervalMs);
    // A pending usage send must never keep the process alive on quit.
    this.flushTimer.unref?.();
  }

  stop(): void {
    if (this.flushTimer !== undefined) clearInterval(this.flushTimer);
    this.flushTimer = undefined;
  }

  /** The switch was turned off. */
  sharingWasTurnedOff(): void {
    this.throwAwayEverythingHeld();
    try {
      this.options.eraseEverythingSent(this.options.installIdentifier());
    } catch {
      // Best effort, like every send.
    }
  }

  /**
   * One tick. Fire-and-forget: returns the in-flight promise only so tests can
   * wait for it; production never awaits it.
   */
  sendWhatIsHeld(): Promise<void> {
    if (!this.options.isSharingEnabled()) {
      this.throwAwayEverythingHeld();
      return Promise.resolve();
    }
    if (this.aSendIsInFlight || this.heldCounts.size === 0) return Promise.resolve();
    const irisVersion = threePartVersion(this.options.irisVersion);
    if (irisVersion === null) {
      this.throwAwayEverythingHeld();
      return Promise.resolve();
    }

    const everythingHeld = [...this.heldCounts.values()];
    const batch = everythingHeld.slice(0, LARGEST_BATCH_THE_SERVER_ACCEPTS);
    // Whatever did not fit is dropped with it: the next minute starts from
    // nothing rather than growing a backlog.
    this.droppedEventCount += everythingHeld.length - batch.length;
    this.heldCounts.clear();

    const events = batch.map((held) => {
      const entry: Record<string, unknown> = {
        event: held.event.kind,
        os: this.options.operatingSystem,
        irisVersion,
        hourBucket: held.hourBucket,
        count: held.count,
      };
      if (held.event.appSlug) entry.appSlug = held.event.appSlug;
      if (held.event.provider) entry.provider = held.event.provider;
      if (held.event.modelTier) entry.modelTier = held.event.modelTier;
      return entry;
    });

    let jsonBody: string;
    try {
      jsonBody = JSON.stringify({ installId: this.options.installIdentifier(), events });
    } catch {
      return Promise.resolve();
    }

    this.aSendIsInFlight = true;
    return this.options
      .sendBatch(jsonBody)
      .catch(() => false)
      .then(() => {
        // Accepted or not, the batch is gone. See the file note.
        this.aSendIsInFlight = false;
      });
  }

  private throwAwayEverythingHeld(): void {
    for (const held of this.heldCounts.values()) this.droppedEventCount += held.count;
    this.heldCounts.clear();
  }

  // For tests.
  heldCountsForTesting(): Array<{ event: UsageEvent; hourBucket: string; count: number }> {
    return [...this.heldCounts.values()].map((held) => ({ ...held }));
  }
  droppedEventCountForTesting(): number {
    return this.droppedEventCount;
  }
  sendIsInFlightForTesting(): boolean {
    return this.aSendIsInFlight;
  }
}

type FetchLike = (
  input: string,
  init?: { method?: string; headers?: Record<string, string>; body?: string; signal?: AbortSignal }
) => Promise<{ ok: boolean }>;

/**
 * The real sender: `POST {publik}/api/telemetry/usage`, no credential of any
 * kind (a usage count has no account), a 10 s timeout, and `false` for any
 * failure instead of a throw.
 */
export function makeUsageSender(publikBaseUrl: string, fetchImplementation: FetchLike): {
  sendBatch: (jsonBody: string) => Promise<boolean>;
  eraseEverythingSent: (installIdentifier: string) => void;
} {
  const endpoint = `${publikBaseUrl.replace(/\/+$/, "")}/api/telemetry/usage`;
  return {
    async sendBatch(jsonBody: string): Promise<boolean> {
      const abort = new AbortController();
      const timeout = setTimeout(() => abort.abort(), 10_000);
      try {
        const response = await fetchImplementation(endpoint, {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: jsonBody,
          signal: abort.signal,
        });
        return response.ok;
      } catch {
        return false;
      } finally {
        clearTimeout(timeout);
      }
    },
    eraseEverythingSent(installIdentifier: string): void {
      void fetchImplementation(`${endpoint}?install_id=${encodeURIComponent(installIdentifier)}`, {
        method: "DELETE",
      }).catch(() => undefined);
    },
  };
}
